jupyterlab_hdf/responses.py

```diff
@@ -0,0 +1,136 @@
+        self._uri = uri
+
+    def metadata(self, **kwargs):
+        return dict(
+            sorted(
+                (
+                    *super().metadata().items(),
+                    ("targetFile", self._target_file),
+                    ("targetUri", self._target_uri),
+                )
+            )
+        )
+
+
+class ResolvedEntityResponse(EntityResponse):
+    def __init__(self, hobj):
+        self._hobj = hobj
+        self._uri = hobj.name
+
+    def attributes(self, attr_keys=None):
+        if attr_keys is None:
+            return dict((*self._hobj.attrs.items(),))
+
+        return dict((key, self._hobj.attrs[key]) for key in attr_keys)
+
+    def metadata(self, **kwargs):
+        attribute_names = sorted(self._hobj.attrs.keys())
+        return dict((*super().metadata().items(), ("attributes", [attrMetaDict(self._hobj.attrs.get_id(k)) for k in attribute_names])))
+
+    @property
+    def uri(self):
+        return self._hobj.name
+
+
+class DatasetResponse(ResolvedEntityResponse):
+    type = "dataset"
+
+    def metadata(self, ixstr=None, min_ndim=None):
+        d = super().metadata()
+        shapekeys = ("labels", "ndim", "shape", "size")
+        smeta = {k: v for k, v in shapemeta(self._hobj.shape, self._hobj.size, ixstr=ixstr, min_ndim=min_ndim).items() if k in shapekeys}
+
+        return dict(
+            sorted(
+                (
+                    ("dtype", self._hobj.dtype.str),
+                    *d.items(),
+                    *smeta.items(),
+                )
+            )
+        )
+
+    def data(self, ixstr=None, subixstr=None, min_ndim=None):
+        return dsetChunk(self._hobj, ixstr=ixstr, subixstr=subixstr, min_ndim=min_ndim)
+
+
+class GroupResponse(ResolvedEntityResponse):
+    type = "group"
+
+    def contents(self, content=False, ixstr=None, min_ndim=None):
+        if not content:
+            return super().contents(ixstr=ixstr, min_ndim=min_ndim)
+
+        # Recurse one level
+        return [
+            create_response(self._hobj, suburi).contents(
+                content=False,
+                ixstr=ixstr,
+                min_ndim=min_ndim,
+            )
+            for suburi in self._hobj.keys()
+        ]
+
+    def metadata(self, **kwargs):
+        d = super().metadata()
+
+        return dict(sorted((*d.items(), ("childrenCount", len(self._hobj)))))
+
+
+def create_response(root: Union[h5py.File, h5py.Group], uri: str):
+    if uri == "/":
+        return GroupResponse(root)
+
+    link = root.get(uri, getlink=True)
+    if isinstance(link, h5py.ExternalLink):
+        root_uri = "" if isinstance(root, h5py.File) else root.name
+        return ExternalLinkResponse(link, uriJoin(root_uri, uri))
+
+    hobj = root[uri]
+    if isinstance(hobj, h5py.Dataset):
+        return DatasetResponse(hobj)
+    elif isinstance(hobj, h5py.Group):
+        return GroupResponse(hobj)
+    else:
+        return ResolvedEntityResponse(hobj)
```

Comment:
The main point of the refactoring: responses are now handled by this hierarchy of class. 

Also, the factory function `create_response` is the only place where the file is accessed. It will therefore be the ideal location if/when we want to resolve the links.